File: output/GPT3.5-Turbo/xCodeEval_4000/C++/C/51d7449e82bf99514d748080a014d4e4.c
#include <stdio.h>
#include <stdlib.h>

#define MAX_N 3010

int n, k;
int m[MAX_N];
int a[MAX_N][MAX_N];
long long ans, sum[MAX_N];

void insert(long long *dp, int a, long long b) {
    for (int i = k; i >= a; i--) {
        dp[i] = dp[i] > dp[i - a] + b ? dp[i] : dp[i - a] + b;
    }
}

void DFS(int l, int r, long long *f) {
    if (l == r) {
        ans = ans > f[k] ? ans : f[k];
        long long s = 0;
        for (int i = 1; i <= m[l] && i <= k; i++) {
            s += a[l][i];
            ans = ans > s + f[k - i] ? ans : s + f[k - i];
        }
        return;
    }
    long long g[MAX_N];
    int mid = (l + r) >> 1;
    for (int i = mid + 1; i <= r; i++) {
        for (int j = k; j >= m[i]; j--) {
            g[j] = g[j] > g[j - m[i]] + sum[i] ? g[j] : g[j - m[i]] + sum[i];
        }
    }
    DFS(l, mid, g);
    for (int i = l; i <= mid; i++) {
        for (int j = k; j >= m[i]; j--) {
            g[j] = g[j] > g[j - m[i]] + sum[i] ? g[j] : g[j - m[i]] + sum[i];
        }
    }
    DFS(mid + 1, r, g);
}

int main() {
    scanf("%d%d", &n, &k);
    for (int i = 1; i <= n; i++) {
        scanf("%d", &m[i]);
        for (int j = 1; j <= m[i]; j++) {
            scanf("%d", &a[i][j]);
            sum[i] += a[i][j];
        }
    }
    DFS(1, n, (long long *)calloc(k + 1, sizeof(long long)));
    printf("%lld\n", ans);
    return 0;
}

File: output/GPT3.5-Turbo/xCodeEval_4000/C++/Java/51d7449e82bf99514d748080a014d4e4.java
import java.util.*;

class Main {
    static int N = 3000 + 10;
    static int n, k;
    static int[] m = new int[N];
    static int[][] a = new int[N][N];
    static long ans;
    static long[] sum = new long[N];

    static void insert(List<Long> dp, int a, long b) {
        for (int i = k; i >= a; i--) {
            dp.set(i, Math.max(dp.get(i), dp.get(i - a) + b));
        }
    }

    static void DFS(int l, int r, List<Long> f) {
        if (l == r) {
            ans = Math.max(ans, f.get(k));
            long s = 0;
            for (int i = 1; i <= m[l] && i <= k; i++) {
                s += a[l][i];
                ans = Math.max(ans, s + f.get(k - i));
            }
            return;
        }
        List<Long> g = new ArrayList<>(f);
        int mid = (l + r) >> 1;
        for (int i = mid + 1; i <= r; i++) {
            insert(g, m[i], sum[i]);
        }
        DFS(l, mid, g);
        g = new ArrayList<>(f);
        for (int i = l; i <= mid; i++) {
            insert(g, m[i], sum[i]);
        }
        DFS(mid + 1, r, g);
    }

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        n = sc.nextInt();
        k = sc.nextInt();
        for (int i = 1; i <= n; i++) {
            m[i] = sc.nextInt();
            for (int j = 1; j <= m[i]; j++) {
                a[i][j] = sc.nextInt();
                sum[i] += a[i][j];
            }
        }
        DFS(1, n, new ArrayList<>(Collections.nCopies(k + 1, 0L)));
        System.out.println(ans);
    }
}
